warehouse/attestations/services.py

```diff
@@ -0,0 +1,227 @@
+    Publisher,
+    VerificationError,
+)
+from pyramid.request import Request
+from sigstore.verify import Verifier
+from zope.interface import implementer
+
+from warehouse.attestations.errors import (
+    AttestationUploadError,
+    UnsupportedPublisherError,
+)
+from warehouse.attestations.interfaces import IReleaseVerificationService
+from warehouse.attestations.models import Attestation as DatabaseAttestation
+from warehouse.metrics.interfaces import IMetricsService
+from warehouse.oidc.models import (
+    GitHubPublisher as GitHubOIDCPublisher,
+    GitLabPublisher as GitLabOIDCPublisher,
+    OIDCPublisher,
+)
+from warehouse.packaging.interfaces import IFileStorage
+from warehouse.packaging.models import File
+
+
+def _publisher_from_oidc_publisher(publisher: OIDCPublisher) -> Publisher:
+    """
+    Convert an OIDCPublisher object in a pypy-attestations Publisher.
+    """
+    match publisher.publisher_name:
+        case "GitLab":
+            publisher = typing.cast(GitLabOIDCPublisher, publisher)
+            return GitLabPublisher(
+                repository=publisher.project_path, environment=publisher.environment
+            )
+        case "GitHub":
+            publisher = typing.cast(GitHubOIDCPublisher, publisher)
+            return GitHubPublisher(
+                repository=publisher.repository,
+                workflow=publisher.workflow_filename,
+                environment=publisher.environment,
+            )
+        case _:
+            raise UnsupportedPublisherError
+
+
+@implementer(IReleaseVerificationService)
+class ReleaseVerificationService:
+
+    def __init__(
+        self,
+        storage: IFileStorage,
+        metrics: IMetricsService,
+    ):
+        self.storage: IFileStorage = storage
+        self.metrics: IMetricsService = metrics
+
+    @classmethod
+    def create_service(cls, _context, request: Request):
+        return cls(
+            storage=request.find_service(IFileStorage),
+            metrics=request.find_service(IMetricsService),
+        )
+
+    def persist_attestations(
+        self, oidc_publisher: OIDCPublisher, attestations: list[Attestation], file: File
+    ) -> None:
+        for attestation in attestations:
+            with tempfile.NamedTemporaryFile() as tmp_file:
+                tmp_file.write(attestation.model_dump_json().encode("utf-8"))
+
+                attestation_digest = hashlib.file_digest(
+                    tmp_file, "blake2b"
+                ).hexdigest()
+                database_attestation = DatabaseAttestation(
+                    file=file, attestation_file_blake2_digest=attestation_digest
+                )
+
+                self.storage.store(
+                    database_attestation.attestation_path,
+                    tmp_file.name,
+                    meta=None,
+                )
+
+            file.attestations.append(database_attestation)
+
+        self.generate_and_store_provenance_file(oidc_publisher, attestations, file)
```

Comment:
I think I'd rather do this wherever we need it after `persist_attestations` is called instead.